pkg/ffi/ffi_param_validator.go

```diff
@@ -0,0 +1,251 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ffi
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/santhosh-tekuri/jsonschema/v5"
+)
+
+type ParamValidator struct{}
+
+var intRegex, _ = regexp.Compile("^u?int([0-9]{1,3})$")
+var bytesRegex, _ = regexp.Compile("^bytes([0-9]{1,2})?")
+
+func (v *ParamValidator) Compile(ctx jsonschema.CompilerContext, m map[string]interface{}) (jsonschema.ExtSchema, error) {
+	valid := true
+	if details, ok := m["details"]; ok {
+		var jsonTypeString string
+		n, _ := details.(map[string]interface{})
+		blockchainType := n["type"].(string)
+		jsonType, ok := m["type"]
+		if ok {
+			jsonTypeString = jsonType.(string)
+		} else {
+			_, ok := m["oneOf"]
+			if ok {
+				jsonTypeString = integerType
+			}
+		}
+		switch jsonTypeString {
+		case stringType:
+			if blockchainType != stringType &&
+				blockchainType != addressType &&
+				!isEthereumNumberType(blockchainType) &&
+				!isEthereumBytesType(blockchainType) {
+				valid = false
+			}
+		case integerType:
+			if !isEthereumNumberType(blockchainType) {
+				valid = false
+			}
+		case booleanType:
+			if blockchainType != boolType {
+				valid = false
+			}
+		case arrayType:
+			if !strings.HasSuffix(blockchainType, "[]") {
```

Comment:
@nguyer - trying to work out what this `blockchainType` is, and where this `[]` suffix comes from?

Also overall what this function is for? What is the `ParamValidator` for, and where is it required?

Reply:
Perhaps some comments are needed in the code to explain this. I can add that, but I'll answer here as well...

The `ParamValidator` is what defines the rules for what makes an FFI Param schema valid or invalid. There is a `ParamValidator` in `firefly-common` as well, which defines the rules to which _all_ FFI Param schemas must adhere. This particular `ParamValidator` defines the rules for an _Ethereum specific_ FFI Param, which is why it's in `firefly-signer` along with the rest of the Ethereum specific FFI code.

This particular function checks things beyond just a schema match, which would otherwise be difficult to do in JSON Schema itself. It's primary function is to check that a given `type` (the JSON input type on an API request) is compatible with the Solidity type (the `blockchainType` field in this code) specified inside the `details` part of the schema. For example, this specific line ensures that if the JSON input type is `array` that the Solidity type is also an array.

Just a point of clarification (because I get confused on this all the time): this code does _not_ validate API request input itself. That's up to the blockchain connector. This is just making sure that the FFI schema that was provided is in fact a valid schema.

Reply:
This is super helpful @nguyer - so I think this code should just be doing a `HasSuffix` check for `]` on the type then.
If I understand correctly, a type like `uint64[][32]` array would also have an array input in the schema. Might be worth a test for a fixed-length array type to confirm it round-trips

Reply:
If I understand correctly, this is just validating the top-level of parameters of the inputs, not the nested ones.

For example, here's an arbitrarily overcomplicated, but valid, Solidity file with a mix of nested types:

```solidity
pragma solidity ^0.8.0;

contract ComplexType {
  uint public storedData;

  struct MyType1 {
    string           strField;
    bytes            bytesField;
    bytes32          someHash;
    bytes32[]        lotsOfHashs;
    bytes32[5]       fiveHashes;
    ufixed256x18[10] tenValues;
    MyType2          innerStruct;
    MyType2[]        innerVarArray;
    MyType2[5][][5]  innerFixedVarArr;
  } 

  struct MyType2 {
    string           strField;
    bytes            bytesField;
    bytes32          someHash;
    bytes32[]        lotsOfHashs;
    bytes32[5]       fiveHashes;
  }

  uint256 val;

  constructor() {}

  function doStuff(MyType1 memory, MyType1[] memory, MyType1[5] memory) public {
    val = val + 1;
  }

}
```

Here you'd see multiple levels of types in the JSON, for each type. I can see how this could would ensure that the top-level `MyType1[]` in the JSON schema is an array type. But per the above I think it would reject `MyType[5]` being an array.

But then what about all the other levels?

Then there's `isEthereumNumberType` - this seems to imply that for a subset (not all - as it doesn't cover `ufixed`) we expect the input type to be number. But I understand we use a `oneOf` there.

Then the connectors do accept different forms for some of these inputs, beyond the ones we cover here. For example the firefly-signer code (as used by EVM Connect) allows arrays for input on tuples, not just objects. So if you were hand crafting your FFI rather than using the generator, you could validly instruct your users to supply an array.

Sorry this is rambling. My concern isn't that we're not covering all edge cases, but rather that we're trying to be restrictive while knowing we haven't covered all edge cases.

I wonder if it would be better to just be less restrictive, and simply allow anything that we can process successfully. 

e.g. if we can run the JSON Schema through the Schema->ABI converter, and get a good result, then it's ok. If they choose to use different types on the JSON Schema that's fine - as long as we have a `details.type` on there, so that we can re-construct the ABI from it.

Reply:
After a good long chat about this, we agreed that we should be using the new `TypeComponentTree` to make the decisions about what the ABI type is here and use that to ensure the JSON input type is a match.